pkg/services/ngalert/accesscontrol/silences.go

```diff
@@ -0,0 +1,252 @@
+	if err != nil || !canGlobal { // if user cannot read global silences, then it does not have silence permissions at all.
+		return nil, err
+	}
+	result := make([]Silence, 0, len(silences))
+	keys := make(map[string][]Silence, len(silences))
+	for _, silence := range silences {
+		ruleUID := silence.GetRuleUID()
+		if ruleUID == nil { // if this is a global silence
+			result = append(result, silence)
+			continue
+		}
+		key := *ruleUID
+		keys[key] = append(keys[key], silence)
+	}
+	if len(keys) == 0 { // if only global silences are provided no need in other checks
+		return result, nil
+	}
+	namespaces, err := s.store.GetNamespacesByRuleUID(ctx, user.GetOrgID(), maps.Keys(keys)...)
+	if err != nil {
+		return nil, err
+	}
+
+	for key, silence := range keys {
+		ns, ok := namespaces[key]
+		if !ok {
+			continue
+		}
+		hasAccess, err := s.HasAccess(ctx, user, readRuleSilenceEvaluator(ns))
+		if err != nil {
+			return nil, err
+		}
+		if !hasAccess {
+			continue
+		}
+		result = append(result, silence...)
+	}
+	return result, nil
+}
+
+// AuthorizeReadSilence checks if user has access to read a silence
+func (s SilenceService) AuthorizeReadSilence(ctx context.Context, user identity.Requester, silence Silence) error {
+	can, err := s.HasAccess(ctx, user, readSomeSilenceEvaluator)
+	if err != nil {
+		return err
+	}
+	if !can { // if user cannot read global silences, then it does not have silence permissions at all.
+		return NewAuthorizationErrorWithPermissions("read any silences", readSomeSilenceEvaluator)
+	}
+	ruleUID := silence.GetRuleUID()
+	if ruleUID == nil {
+		return nil // user can read global
+	}
+
+	canAll, err := s.HasAccess(ctx, user, readAnySilenceEvaluator)
+	if canAll || err != nil { // exit if user can either read all silences or there is error
+		return err
+	}
+
+	// otherwise resolve rule key to the action's scope
+	folderUID, err := s.ruleUIDToFolderUID(ctx, user.GetOrgID(), *ruleUID)
+	if err != nil {
+		return fmt.Errorf("resolve rule UID to folder UID: %w", err)
+	}
+	if folderUID == "" { // if we did not find folder by rule UID then it does not exist.
+		return NewAuthorizationErrorGeneric(fmt.Sprintf("read silence for rule %s", *ruleUID))
+	}
+	return s.HasAccessOrError(ctx, user, readRuleSilenceEvaluator(folderUID), func() string {
+		return "read silence"
+	})
+}
+
+// AuthorizeCreateSilence checks if user has access to create a silence. Returns ErrAuthorizationBase if user is not authorized
+func (s SilenceService) AuthorizeCreateSilence(ctx context.Context, user identity.Requester, silence Silence) error {
+	ruleUID := silence.GetRuleUID()
+	if ruleUID == nil {
+		return s.HasAccessOrError(ctx, user, createGlobalSilenceEvaluator, func() string {
+			return "create silences that can potentially affect multiple rules"
+		})
+	} else {
+		canAll, err := s.HasAccess(ctx, user, createAnySilenceEvaluator)
+		if err != nil || canAll { // exit if user can either read all silences or there is error
+			return err
```

Comment:
Nit: Probably good to indicate `err` can be nil here in the second case, to make it obvious to anyone coming back to this later.